pretyping/pretyping.ml

```diff
@@ -662,20 +683,29 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
         let p = Projection.make p false in
         let npars = Projection.npars p in
         fun n ->
-          if n == npars + 1 then fun _ v -> mkProj (p, v)
+          if Int.equal n npars then fun _ v -> mkProj (p, v)
           else fun f v -> applist (f, [v])
       | _ -> fun _ f v -> applist (f, [v])
     in
-    let rec apply_rec env sigma n resj candargs = function
-      | [] -> sigma, resj
+    let rec apply_rec env sigma n resj candargs bidiargs = function
+      | [] -> sigma, resj, List.rev bidiargs
       | c::rest ->
+        let bidi = n >= nargs_before_bidi in
         let argloc = loc_of_glob_constr c in
         let sigma, resj = Coercion.inh_app_fun ~program_mode resolve_tc !!env sigma resj in
         let resty = whd_all !!env sigma resj.uj_type in
         match EConstr.kind sigma resty with
         | Prod (na,c1,c2) ->
           let tycon = Some c1 in
-          let sigma, hj = pretype tycon env sigma c in
+          let (sigma, hj), bidiargs =
+            if bidi then
```

Comment:
Is it useful to do the bidi stuff when there's no tycon? Otherwise we may add `&& tycon <> None` to this condition.